File: backend/app/common/enums.py
"""Shared enums (api-specification.md §4.5 / database-schema.md §1.5).

These are str-valued Python enums whose VALUES are the exact lowercase wire labels
AND the exact Postgres native enum labels. `PG_ENUM_NAMES` maps each enum to its
Postgres native type name (schema §1.5), used by the SQLAlchemy column + the
migration's explicit `create_type`.

Letter grades are intentionally NOT an enum — they derive on read from the
configurable grading_scale_bands (D11) and travel the wire as plain strings.
"""

from __future__ import annotations

import enum


class Role(str, enum.Enum):
    """One role per user (A-ONE-ROLE). Values are the wire labels AND the DB enum labels.

    The DISPLAY vocabulary differs and is a frontend concern (D30, decision #3):
    principal -> Dean, secretary -> Registrar, teacher -> Lecturer. The values below were
    deliberately not renamed with it.

    D43 added the last two:

      AUDITOR  Reads everything, writes nothing. Not "an admin who should behave" - the
               refusal is enforced in `get_current_user`, which every authenticated route
               passes through, so no route can opt out of it by omission. `require_role`
               is a role allowlist and cannot express "GET only", which is why the guard
               lives there and not in 121 dependency tuples.

      HOD      Head of Department: a LECTURER who also supervises a programme. Keeps every
               lecturer power on offerings they are actually assigned to - ownership, not
               role, is what limits that - and additionally READS everything in the
               programme(s) they head. The link is `program_heads`; there is no
               `departments` table and a programme is the unit BAJC actually has.
    """

    PRINCIPAL = "principal"
    SECRETARY = "secretary"
    TEACHER = "teacher"
    STUDENT = "student"
    HOD = "hod"
    AUDITOR = "auditor"


#: Roles that may never write, whatever route they reach. Enforced centrally in
#: `app.core.deps.get_current_user` - see `Role.AUDITOR` above.
READ_ONLY_ROLES: frozenset[Role] = frozenset({Role.AUDITOR})

#: Roles that carry a `teacher_profiles` row and can therefore own an offering.
#: Anywhere that resolves a lecturer profile from the principal must accept both, or an
#: HOD silently becomes a lecturer with no offerings rather than one with their own.
LECTURER_ROLES: frozenset[Role] = frozenset({Role.TEACHER, Role.HOD})


class StudentStatus(str, enum.Enum):
    """Student lifecycle, in the CLIENT'S vocabulary (D34).

    Adopted verbatim from the client's own `student_profiles` dump, including its comment
    on what each state means:

      REGISTERED    "instead of active" - enrolled and attending.
      UNREGISTERED  "when student do not continue further semesters, but has successfully
                    completed the last semester". NOT a failure state, which is why
                    "inactive" was the wrong word for it.
      DROPOUT       left mid-programme. Pairs with `dropout_date` / `dropout_reason`.
      GRADUATED     "Dean/Registrar are the only ones with access to change to this
                    status" - already true, since POST /students/{id}/status is
                    require_role(PRINCIPAL, SECRETARY).
      TRANSFERRED / WITHDRAWN  unchanged.

    **The mixed case is deliberate.** Three values are TitleCase and three are lowercase
    because that is exactly how the client's enum reads, and their dump is the authority
    for this column - normalising it would put the database out of step with their own
    tooling for a cosmetic gain. MariaDB's `utf8mb4_uca1400_ai_ci` is case-insensitive, so
    comparison is unaffected; the case matters only for what is stored and echoed.

    The MEMBER names track the values (`REGISTERED`, not `ACTIVE`) so a reader of
    `StudentStatus.REGISTERED` sees the word the registry actually uses. That rename is
    why D34 touches ~20 call sites it otherwise would not have.
    """

    REGISTERED = "Registered"
    UNREGISTERED = "Unregistered"
    DROPOUT = "DropOut"
    TRANSFERRED = "transferred"
    GRADUATED = "graduated"
    WITHDRAWN = "withdrawn"


class TeacherStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"


class AcademicYearStatus(str, enum.Enum):
    ACTIVE = "active"
    ARCHIVED = "archived"


class TermType(str, enum.Enum):
    """The KIND of calendar term (D30 §D3, `semesters.term_type`).

    BAJC does not run two symmetrical semesters a year: its programme sequences move
    through Summer and Spring blocks as well as numbered semesters. Before D30,
    `semesters.sequence` carried `CHECK (sequence IN (1,2))` and the school could not
    record more than two terms in a year at all.

    This is the CALENDAR term's kind. A course's position in a programme PLAN is a
    different fact and lives on `program_courses.term_label` — "Semester 1" there means
    "the first semester of this plan", not any particular dated term (§D3).
    """

    SUMMER = "summer"
    SEMESTER = "semester"
    SPRING = "spring"


class CourseComponent(str, enum.Enum):
    """BAJC's classification of a catalog course (D30 §D2, plan §A1).

    Upper-case values because that is how they are written on the programme course
    sequences and on `courses.component enum('GEC','SEC','CEC')` — the one place in
    this file where the wire value is not lower-case, and it is the source documents
    that decide that, not a style preference.
    """

    GEC = "GEC"  # General Education Core
    SEC = "SEC"  # Supporting Education Core
    CEC = "CEC"  # Concentration Education Core


class PrerequisiteType(str, enum.Enum):
    """What kind of requirement a `course_prerequisites` row expresses (D30 §D4).

    `ALL_PROGRAM_COURSES` exists for one real case in the source material, and it is
    not an edge case — it is `EDUC3201` (Internship), whose prerequisite is written
    on the Primary Education sequence as literally **"ALL COURSES"**. A list of
    course ids cannot say that: the gate has to keep meaning "everything in the
    programme" as the programme's curriculum changes.
    """

    COURSE = "course"
    ALL_PROGRAM_COURSES = "all_program_courses"


class Gender(str, enum.Enum):
    """The gender values the forms offer (D37).

    **A `str` enum, but the COLUMNS stay free text**, and that split is deliberate.
    `student_profiles.gender` and `applications.gender` are `varchar` with the comment
    "Free/lookup text; not a fixed enum" — narrowing them to a DB enum would reject the
    historical rows this system did not write, and there is no migration that can safely
    guess what a value it has never seen was meant to be.

    So this constrains the WRITE PATH instead: the forms offer exactly these two, and
    `normalise_gender` folds anything else onto them before it is stored. Reads stay
    permissive.

    Lowercase because that is what 45 of the 47 live rows already used, and what every
    frontend comparison (`gender === 'female'`) and the demo dataset assume.

    **The drift was on BOTH tables, and a `GROUP BY gender` could not show it.** Under
    `utf8mb4_uca1400_ai_ci` the collation is case-insensitive, so `'Male'` and `'male'`
    collapse into one group and the query reports whichever it saw first. It took
    `GROUP BY HEX(gender)` to see that `student_profiles` held one `'Male'` and
    `applications` held another. Both were folded to lowercase in D37; `normalise_gender`
    is what stops them coming back.
    """

    FEMALE = "female"
    MALE = "male"


#: Accepted spellings -> the canonical value. Deliberately generous: a Registrar
#: transcribing a paper form, an import, and the client's own dump have all produced
#: different capitalisations, and rejecting them would block a legitimate record over a
#: letter case.
_GENDER_ALIASES: dict[str, Gender] = {
    "f": Gender.FEMALE,
    "female": Gender.FEMALE,
    "woman": Gender.FEMALE,
    "girl": Gender.FEMALE,
    "m": Gender.MALE,
    "male": Gender.MALE,
    "man": Gender.MALE,
    "boy": Gender.MALE,
}


def normalise_gender(value: str | None) -> str | None:
    """Fold a submitted gender onto the canonical vocabulary.

    `None` and blank pass through as `None` — gender is optional on both forms and an
    empty string is not a value. **An UNRECOGNISED value is returned unchanged**, not
    rejected: this runs on every write, including the admissions transcription path, and
    turning an unexpected spelling into a 422 would stop a Registrar recording a real
    student over something cosmetic. The dropdowns are what keep new data clean; this is
    the safety net under them.

    Case- and whitespace-insensitive, because that is the drift actually observed —
    `'Male'` on `applications` beside `'male'` on `student_profiles`.

    **Returns a plain `str`, never the enum member.** `Gender` subclasses `str`, but from
    Python 3.11 `str(Gender.MALE)` is `'Gender.MALE'` — so handing the member to a plain
    `varchar` column (these two are free text, not `enum_col`) risks storing that literal
    the moment anything in the driver stringifies it. `.value` removes the question.
    """
    if value is None:
        return None
    cleaned = value.strip()
    if not cleaned:
        return None
    match = _GENDER_ALIASES.get(cleaned.casefold())
    return match.value if match is not None else cleaned


class CivilStatus(str, enum.Enum):
    """Civil status, the values the forms offer (D40, client ask).

    **A `str` enum, but the COLUMNS stay free text** - the same split D37 settled for
    `Gender`, for the same reason. `student_profiles.civil_status` and
    `applications.civil_status` are `varchar(50)`; narrowing them to a DB enum would reject
    the historical rows this system did not write, and no migration can safely guess what a
    value it has never seen was meant to be.

    So this constrains the WRITE PATH: the dropdowns offer exactly these four, and
    `normalise_civil_status` folds recognised spellings onto them before they are stored.
    Reads stay permissive.

    TitleCase, unlike `Gender`, because that is what the live dump already holds
    (`student_profiles.civil_status` = 'Single'). Matching the data beats matching the
    other enum's style - re-casing it would put this system out of step with the client's
    own tooling for a cosmetic gain, and MariaDB's case-insensitive collation means the
    database cannot see the difference anyway. The BROWSER can, which is the whole reason
    the normaliser exists: a stray 'single' renders an empty select and the next save
    writes NULL over a real value.

    `WIDOWER` keeps its parenthetical because the client's paper form does. It is one
    status, not two; splitting it would make the dropdown ask for a fact the form does not
    collect.
    """

    SINGLE = "Single"
    MARRIED = "Married"
    DIVORCED = "Divorced"
    WIDOWER = "Widow(er)"


#: Accepted spellings -> the canonical value. Generous for the same reason
#: `_GENDER_ALIASES` is: a Registrar transcribing a paper form, an import, and the
#: client's own dump have each produced a different spelling of the same status, and
#: rejecting one over a letter case would block a real record.
#:
#: `widow` and `widower` both map to `Widow(er)` - the parentheses are unlikely to survive
#: a hand-typed import, and the form does not distinguish them.
_CIVIL_STATUS_ALIASES: dict[str, CivilStatus] = {
    "single": CivilStatus.SINGLE,
    "s": CivilStatus.SINGLE,
    "married": CivilStatus.MARRIED,
    "m": CivilStatus.MARRIED,
    "divorced": CivilStatus.DIVORCED,
    "d": CivilStatus.DIVORCED,
    "widow(er)": CivilStatus.WIDOWER,
    "widow": CivilStatus.WIDOWER,
    "widower": CivilStatus.WIDOWER,
    "widowed": CivilStatus.WIDOWER,
    "w": CivilStatus.WIDOWER,
}


def normalise_civil_status(value: str | None) -> str | None:
    """Fold a submitted civil status onto the canonical vocabulary.

    Behaves exactly like `normalise_gender`, and deliberately so - one rule for both
    free-text vocabularies is one rule to remember:

    * `None` and blank pass through as `None`. The field is optional on the column and an
      empty string is not a value.
    * **An UNRECOGNISED value is returned unchanged**, not rejected. This runs on every
      write including the admissions transcription path, and turning an unexpected spelling
      ('Common law') into a 422 would stop a Registrar recording a real student over
      something cosmetic. The dropdowns keep new data clean; this is the safety net.
    * Case- and whitespace-insensitive, which is the drift that actually occurs.

    Returns a plain `str`, never the enum member - see `normalise_gender` for why
    `.value` is not optional here.
    """
    if value is None:
        return None
    cleaned = value.strip()
    if not cleaned:
        return None
    match = _CIVIL_STATUS_ALIASES.get(cleaned.casefold())
    return match.value if match is not None else cleaned


class EnrollmentStatus(str, enum.Enum):
    """How a student is sitting one course offering (D30 §D2 step 4).

    `sims_bk.sql` carried this as `courses.coursestatus` — an ENROLMENT fact parked
    on the CATALOG, where it would have applied to every student taking the course at
    once. `005_tertiary.sql` §8 moved it to `class_enrollments` and added `enrolled`,
    which the original enum lacked entirely despite being the normal case.
    """

    ENROLLED = "enrolled"
    AUDIT = "audit"
    WITHDRAW_PASSING = "withdraw_passing"
    WITHDRAW_FAILING = "withdraw_failing"


class AssessmentType(str, enum.Enum):
    QUIZ = "quiz"
    TEST = "test"
    EXAM = "exam"
    ASSIGNMENT = "assignment"


class AssessmentStatus(str, enum.Enum):
    DRAFT = "draft"
    PUBLISHED = "published"
    GRADING = "grading"
    GRADED = "graded"


class GradeStatus(str, enum.Enum):
    PENDING = "pending"
    GRADED = "graded"
    ABSENT = "absent"
    EXCUSED = "excused"
    EXEMPT = "exempt"


class AttendanceStatus(str, enum.Enum):
    PRESENT = "present"
    ABSENT = "absent"
    LATE = "late"
    EXCUSED = "excused"


class AnnouncementAudience(str, enum.Enum):
    ALL = "all"
    STUDENTS = "students"
    TEACHERS = "teachers"
    CLASS = "class"


class DayOfWeek(int, enum.Enum):
    """ISO weekday of a class meeting (D29 timetable).

    Deliberately NOT in `PG_ENUM_NAMES`: this one is stored as a plain SmallInteger
    rather than a native enum so `ORDER BY day_of_week, start_time` yields Mon→Fri
    order for the timetable grid (a native enum would sort by label text). The DB
    CHECK is `BETWEEN 1 AND 5` — sixth-form timetables are weekday-only, so there is
    no SATURDAY/SUNDAY member; ISO numbering leaves room to relax that later without
    renumbering the weekdays.
    """

    MONDAY = 1
    TUESDAY = 2
    WEDNESDAY = 3
    THURSDAY = 4
    FRIDAY = 5


# ── Admissions (D30 §D11, Phase 4; DDL in `005_tertiary.sql` §11–§14) ──────────
class ApplicationStatus(str, enum.Enum):
    """Where an admission application has got to (`applications.status`).

    This is NOT `student_profiles.status`, and conflating the two is the mistake the
    audit found in `sims_bk.sql`: that dump had no applicant entity at all, so the
    official-use block of the paper form had nowhere to go and the student lifecycle
    enum was being asked to double as an admission decision (plan §B4).

    `DRAFT` is a real state, not a convenience. The Registrar transcribes a paper form
    section by section, so a half-entered application has to survive being interrupted
    — and only a `SUBMITTED` one is a decision waiting to be made.

    `WITHDRAWN` is the applicant pulling out, as distinct from `DENIED`, the college
    saying no. Both are terminal and neither is a delete: an admissions record is kept.
    """

    DRAFT = "draft"
    SUBMITTED = "submitted"
    UNDER_REVIEW = "under_review"
    ACCEPTED = "accepted"
    DENIED = "denied"
    WITHDRAWN = "withdrawn"


class District(str, enum.Enum):
    """Belize's six districts (`applications.district`, `student_profiles.district`).

    Title case because that is how they are written on the application form and in the
    column — the same reasoning as `CourseComponent`: the source document decides the
    wire value, not a style preference.
    """

    COROZAL = "Corozal"
    ORANGE_WALK = "Orange Walk"
    BELIZE = "Belize"
    CAYO = "Cayo"
    STANN_CREEK = "Stann Creek"
    TOLEDO = "Toledo"


class YearOfStudy(str, enum.Enum):
    """First or Second year (`applications.year_of_study`).

    **Half of a field the source dump conflated.** `sims_bk.sql` carried
    `yearofstudy enum('Summer','First','Second','Part Time','Full Time','Transient')`,
    which mixed the applicant's YEAR with their study LOAD — two independent questions
    that Section E of the form asks separately, and which a single column cannot answer
    both of. `005` §9 split them; this is the year half and `EnrollmentLoad` is the
    other.
    """

    FIRST = "First"
    SECOND = "Second"


class EnrollmentLoad(str, enum.Enum):
    """Study load (`applications.enrollment_load`) — the other half of the split above.

    The form defines the first two by credits: Part Time is under 15 credits a term,
    Full Time is over 15. `TRANSIENT` is a visiting student taking courses without
    reading for a BAJC award. The thresholds are NOT enforced here — a load is what the
    applicant declares at admission, while their actual credits are a consequence of
    enrolment that changes term by term.
    """

    PART_TIME = "Part Time"
    FULL_TIME = "Full Time"
    TRANSIENT = "Transient"
    #: D34 — from the client's `yearofstudy` enum, which conflated the year with the load.
    #: 'Summer' names a TERM rather than a load, but it answers the same question this
    #: column does ("how is this student attending") and it goes here rather than on
    #: `YearOfStudy` because a summer student still has a first or second year — putting it
    #: there would make that unanswerable, which is the exact defect the D30 split fixed.
    SUMMER = "Summer"


class EducationLevel(str, enum.Enum):
    """The level of a prior institution (`application_education.education_level`).

    Section B of the form offers exactly these two. The distinction is load-bearing for
    credit transfer: policy allows transfer only from a recognised TERTIARY institution
    (brief §13), so a high-school row can never support one.
    """

    HIGH_SCHOOL = "High School"
    TERTIARY = "Tertiary"


class ApplicationDocumentType(str, enum.Enum):
    """The admission document checklist (`application_documents.document_type`).

    The first five are Section F of the form verbatim. The last three exist for credit
    transfer, whose policy requires a CTA, an original transcript and course outlines
    (brief §13) — `credit_transfer_requests` FKs its three document columns here.

    There is deliberately no file upload behind these rows: object storage is not
    provisioned in this repo (TODO(OQ-DB5), the same reason the school-logo upload is
    still a stub), and Section F on paper IS a tick-list. `received` is the field that
    matters; `file_name` / `content_type` / `size_bytes` are metadata for when storage
    lands, and adding real bytes later is additive.
    """

    PASSPORT_PHOTO = "passport_photo"
    HS_DIPLOMA = "hs_diploma"
    RECOMMENDATION_FORM = "recommendation_form"
    SOCIAL_SECURITY_CARD = "social_security_card"
    COURSE_OUTLINE = "course_outline"
    TRANSCRIPT = "transcript"
    CTA = "cta"
    OTHER = "other"


class GradeRevisionStatus(str, enum.Enum):
    """A grade-revision request's decision state (D30 §D7, brief §20).

    **The workflow, not the score.** `assessment_grades.makeup_score` plus the
    `allow_makeup` policy chain already carried the second-attempt SCORE before D30; what
    was missing was the request, the reason, the approval and the history (plan §B2). This
    enum is the approval half.

    A `PENDING` row is what makes the Dean's queue a filtered read rather than a new
    notifications table (§D8), and `grade_revision_requests.pending_flag` — generated as
    `IF(status = 'pending', 1, NULL)` under a unique index — is what limits a grade to ONE
    open request at a time.

    There is no `withdrawn`: a Lecturer who changes their mind deletes the pending request,
    because an un-ruled request carries no decision worth keeping. A DECIDED one is never
    deleted and never re-opened — the ruling is the record.
    """

    PENDING = "pending"
    APPROVED = "approved"
    DENIED = "denied"


class CreditTransferStatus(str, enum.Enum):
    """A credit-transfer request's decision state (`credit_transfer_requests.status`).

    **The Dean decides** (brief §13, plan §D14), and approval is the only transition the
    database itself polices: `ck_cta_approval_requires_75` refuses an `approved` row
    whose `content_equivalency_pct` is null or below 75. The floor is checked on
    APPROVAL rather than on creation on purpose — a request may legitimately be filed
    before anyone has assessed the equivalency.
    """

    PENDING = "pending"
    APPROVED = "approved"
    DENIED = "denied"


class ReportCardKind(str, enum.Enum):
    """Which report card a `report_card_snapshots` row holds (D32, brief §5).

    **The two are produced by different mechanisms, not just at different times**, and
    that is why they need distinguishing rather than a date range:

      * `MIDTERM` is a FROZEN document. It is captured once when the term's mid-term
        window closes and is served back verbatim thereafter — the client's requirement
        that a mid-term report must not recalculate from current grades.
      * `ENDTERM` is COMPUTED ON READ while the year is live, and only becomes a frozen
        row when the year archives. Its figures are supposed to keep moving until then.

    A term therefore holds at most one of each per student, which is what
    `uq_report_card_snapshot (student_id, semester_id, kind)` enforces.
    """

    MIDTERM = "midterm"
    ENDTERM = "endterm"


# Maps each Python enum to its Postgres native enum type name (schema §1.5).
# The migration creates these types explicitly (create_type) before any table
# uses them; the model columns reference the same names with create_type=False.
PG_ENUM_NAMES: dict[type[enum.Enum], str] = {
    Role: "user_role",
    StudentStatus: "student_status",
    TeacherStatus: "teacher_status",
    AcademicYearStatus: "academic_year_status",
    AssessmentType: "assessment_type",
    AssessmentStatus: "assessment_status",
    GradeStatus: "grade_status",
    AttendanceStatus: "attendance_status",
    AnnouncementAudience: "announcement_audience",
}

File: frontend/src/shared/types/enums.ts
/**
 * Enums mirroring the Postgres native enums (api-specification.md §4.5).
 *
 * `Role` is now RE-EXPORTED from the orval-generated client (the auth OpenAPI
 * defines it), so the role vocabulary cannot drift from the backend. The remaining
 * enums below are still hand-authored — their owning modules (Students, Assessments,
 * …) are not yet in the served OpenAPI; they will be re-sourced from codegen as those
 * endpoints land. All are string-literal unions (not TS `enum`s) so they serialize/
 * compare to the exact lowercase wire labels with zero runtime cost.
 */

// Generated role vocabulary (single source of truth). Re-exported as a type so all
// existing `import type { Role } from '@shared/types/enums'` sites keep working.
export type { Role } from '@shared/api/generated/model';
import type { Role } from '@shared/api/generated/model';

/**
 * The gender values the forms offer (D37). Mirrors `app/common/enums.py::Gender`.
 *
 * **The COLUMNS are free text and stay so** — `student_profiles.gender` and
 * `applications.gender` are `varchar` carrying the comment "Free/lookup text; not a fixed
 * enum", because narrowing them would reject historical rows this system did not write.
 * What is constrained is the WRITE PATH: these two are the only values the dropdowns
 * offer, and the server folds anything else onto them (`normalise_gender`).
 *
 * Lowercase because 46 of the 47 live rows already are, and because every comparison in
 * this codebase is written that way — `gender === 'female'`. That matters more than it
 * looks: MariaDB's collation is case-insensitive so the directory FILTER tolerated a
 * stray `'Male'`, but JS does not, and a mismatched value renders an EMPTY select and the
 * wrong label on the profile card.
 */
export type Gender = 'female' | 'male';

export const GENDERS: readonly Gender[] = ['female', 'male'];

/** Display labels. The stored values are lowercase; these are what a human reads. */
export const GENDER_LABEL: Record<Gender, string> = {
  female: 'Female',
  male: 'Male',
};

/**
 * The value as one of the two canonical options, or `null` if it is neither.
 *
 * Used to seed a `<select>`: a value that is not among the options renders BLANK, and
 * saving from a blank select clears the field. Canonicalising on load turns a stored
 * `'Male'` into the Male option instead.
 */
export function canonicalGender(value: string | null | undefined): Gender | null {
  if (!value) return null;
  const key = value.trim().toLowerCase();
  return key === 'female' || key === 'male' ? key : null;
}

/**
 * A stored value as a label, tolerating anything the dropdowns did not write.
 */
export function genderLabel(value: string | null | undefined): string | null {
  if (!value) return null;
  const key = value.trim().toLowerCase();
  if (key === 'female' || key === 'male') return GENDER_LABEL[key];
  // Not one of ours — show it as stored rather than mislabelling it.
  return value;
}

/**
 * Student lifecycle — the blueprint's §7.5 vocabulary (D45, client decision C5).
 *
 * **REPLACES the D34 six.** D34 adopted the client's own dump verbatim, mixed case and
 * all; the revised blueprint lists ten states and BAJC confirmed on 2026-09-08 that the
 * ten are what they want. The old values map on cleanly:
 *
 *   Registered -> Active      Unregistered -> Inactive     DropOut -> Dropout
 *   graduated  -> Graduated   withdrawn    -> Withdrawn    transferred -> Transferred
 *
 * `Transferred` is an ELEVENTH value and is NOT in the blueprint. The ten have no
 * equivalent, one live row carries it, and folding it into `Withdrawn` would have
 * rewritten that student's history to say something untrue about why they left — so the
 * client kept it.
 *
 * ⚠️ **TitleCase, all of them, and the case is load-bearing HERE in a way it is not on
 * the server.** MariaDB's collation is case-insensitive so `status = 'graduated'` matched
 * either spelling; JavaScript's `===` does not. A stray lowercase value renders a BLANK
 * select and the wrong label on the profile card — the same trap `canonicalGender` above
 * documents, and the reason D45 normalised the stored bytes rather than leaving the
 * mixed-case vocabulary in place.
 *
 * Mirrors `app/common/enums.py::StudentStatus`.
 */
export type StudentStatus =
  | 'Applicant'
  | 'Accepted'
  | 'Active'
  | 'Inactive'
  | 'Suspended'
  | 'Withdrawn'
  | 'Dropout'
  | 'Completed'
  | 'Graduated'
  | 'Alumni'
  | 'Transferred';

/** All student statuses, in LIFECYCLE order — which is the order a human reads them. */
export const STUDENT_STATUSES: readonly StudentStatus[] = [
  'Applicant',
  'Accepted',
  'Active',
  'Inactive',
  'Suspended',
  'Completed',
  'Graduated',
  'Alumni',
  'Withdrawn',
  'Dropout',
  'Transferred',
];

export type TeacherStatus = 'active' | 'inactive';

export type AcademicYearStatus = 'active' | 'archived';

export type AssessmentType = 'quiz' | 'test' | 'exam' | 'assignment';

export type AssessmentStatus = 'draft' | 'published' | 'grading' | 'graded';

export type GradeStatus = 'pending' | 'graded' | 'absent' | 'excused' | 'exempt';

export type AttendanceStatus = 'present' | 'absent' | 'late' | 'excused';

export type AnnouncementAudience = 'all' | 'students' | 'teachers' | 'class';

/** All roles, ordered for display/iteration. */
export const ROLES: readonly Role[] = [
  'principal',
  'secretary',
  'teacher',
  'student',
  'hod',
  'auditor',
  'sysadmin',
];

/**
 * Belize's six districts (`app/common/enums.District`).
 *
 * D33 — MOVED HERE from `features/admissions/types.ts`, which re-exports it so every
 * existing import keeps working. It stopped being an admissions-only vocabulary the moment
 * the student record grew an address: two independent copies of a closed enum is exactly
 * how "Stann Creek" ends up spelled two ways in one database.
 */
export type District =
  | 'Corozal'
  | 'Orange Walk'
  | 'Belize'
  | 'Cayo'
  | 'Stann Creek'
  | 'Toledo';

export const DISTRICTS: readonly District[] = [
  'Corozal',
  'Orange Walk',
  'Belize',
  'Cayo',
  'Stann Creek',
  'Toledo',
];

/**
 * Study load (`app/common/enums.EnrollmentLoad`). Part Time is under 15 credits a term,
 * Full Time over 15; Transient is a visiting student. Moved here with `District`, and
 * SEPARATE from year of study — `sims_bk.sql` had one column conflating the two, which
 * could answer neither question (D30 §D11).
 */
export type EnrollmentLoad = 'Part Time' | 'Full Time' | 'Transient' | 'Summer';

export const ENROLLMENT_LOADS: readonly EnrollmentLoad[] = [
  'Part Time',
  'Full Time',
  'Transient',
  // D34 — from the client's `yearofstudy` enum, which conflated the year with the load.
  // 'Summer' names a term, but it answers the same question this field does; putting it on
  // `year_of_study` would make "which year are they in" unanswerable for a summer student.
  'Summer',
];

/**
 * Civil status — the values the forms offer (D40, client ask). Mirrors
 * `app/common/enums.py::CivilStatus`.
 *
 * **The COLUMNS stay free text**, exactly as `Gender` does: `student_profiles.civil_status`
 * and `applications.civil_status` are `varchar(50)`, and narrowing them to a DB enum would
 * reject historical rows this system did not write. What is constrained is the WRITE PATH —
 * these four are the only values the dropdowns offer, and the server folds recognised
 * spellings onto them (`normalise_civil_status`). That is D37's rule, applied to a second
 * field that was still free text.
 *
 * TitleCase, unlike `Gender`, because that is what the live `student_profiles` dump already
 * holds ('Single'). Matching the data beats matching the other enum's style: re-casing it
 * would put this system out of step with the client's own tooling for nothing.
 *
 * `Widow(er)` carries its parenthetical because the client's paper form does; it is one
 * status, not two, and splitting it would make the dropdown ask for a fact the form
 * does not collect.
 */
export type CivilStatus = 'Single' | 'Married' | 'Divorced' | 'Widow(er)';

export const CIVIL_STATUSES: readonly CivilStatus[] = [
  'Single',
  'Married',
  'Divorced',
  'Widow(er)',
];

/**
 * Accepted spellings → the canonical value. Mirrors `_CIVIL_STATUS_ALIASES` in
 * `app/common/enums.py`, entry for entry.
 *
 * Generous on purpose, exactly as the server's is: a Registrar transcribing a paper form,
 * an import, and the client's own dump have each produced a different spelling of the same
 * status, and rejecting one over a letter case would block a real record.
 */
const CIVIL_STATUS_ALIASES: Record<string, CivilStatus> = {
  single: 'Single',
  s: 'Single',
  married: 'Married',
  m: 'Married',
  divorced: 'Divorced',
  d: 'Divorced',
  'widow(er)': 'Widow(er)',
  widow: 'Widow(er)',
  widower: 'Widow(er)',
  widowed: 'Widow(er)',
  w: 'Widow(er)',
};

/**
 * A stored civil status as one of the four canonical options, or `null` if it is none of
 * them. Case- and whitespace-insensitive, and alias-aware: `'widower'` resolves to
 * `Widow(er)` because that is what the server would have folded it to on its next write.
 *
 * Used to SEED a `<select>`, and it is not cosmetic. MariaDB's collation is
 * case-insensitive so a stored `'single'` is indistinguishable from `'Single'` to the
 * database, but JS compares case-sensitively: fed to the select raw it matches no option,
 * renders BLANK, and the next save writes an empty civil status over a real one.
 * Canonicalising on load turns it into the Single option instead — the same fix
 * `canonicalGender` makes for the other free-text vocabulary.
 *
 * **It shares its alias table with `normaliseCivilStatus`**, and that is the point: when
 * the two disagreed, a stored `'widower'` opened as its own "(as recorded)" option while
 * the server considered it `Widow(er)` — one value shown as two, in the one control whose
 * job is to make sure it is one.
 */
export function canonicalCivilStatus(value: string | null | undefined): CivilStatus | null {
  if (!value) return null;
  return CIVIL_STATUS_ALIASES[value.trim().toLowerCase()] ?? null;
}

/**
 * The civil-status options to render, with `current` guaranteed present.
 *
 * The same shape and the same reasoning as `religionOptions`: a stored value absent from
 * the list renders an EMPTY select, and saving from a blank select clears the field — so a
 * legacy 'Common law' would be silently destroyed by an edit that never touched it. It is
 * carried as an extra "(as recorded)" option instead.
 *
 * The known-check is case-SENSITIVE on purpose, so this stays correct even for a caller
 * that did not run `canonicalCivilStatus` first: a raw `'single'` gets its own option
 * rather than matching `'Single'` and then rendering blank against it. Callers that DO
 * canonicalise never reach that branch.
 */
export function civilStatusOptions(
  current: string | null | undefined,
): Array<{ value: string; label: string; legacy: boolean }> {
  const list = CIVIL_STATUSES.map((value) => ({ value, label: value as string, legacy: false }));
  const stored = (current ?? '').trim();
  if (!stored) return list;
  const known = list.some((o) => o.value === stored);
  return known ? list : [...list, { value: stored, label: `${stored} (as recorded)`, legacy: true }];
}

/**
 * Fold a submitted civil status onto the canonical vocabulary — the browser mirror of
 * `app/common/enums.py::normalise_civil_status`.
 *
 * It exists for DEMO MODE. The MSW handlers stand in for the server, and the last two
 * times a write rule lived in only one of the two implementations, demo mode certified a
 * screen the real backend refused. Same contract as the server's:
 *
 *  * blank → `null` (the field is optional and `''` is not a value);
 *  * an UNRECOGNISED value passes through unchanged, never rejected — a Registrar
 *    transcribing 'Common law' from a paper form must not be blocked;
 *  * case- and whitespace-insensitive, which is the drift that actually occurs.
 *
 * The difference from `canonicalCivilStatus` is only what happens to an unrecognised
 * value: this one keeps it (it is a WRITE, and the value is real), that one reports `null`
 * (it is seeding a `<select>`, and the caller needs to know to add an option for it).
 */
export function normaliseCivilStatus(value: string | null | undefined): string | null {
  if (value == null) return null;
  const cleaned = value.trim();
  if (!cleaned) return null;
  return CIVIL_STATUS_ALIASES[cleaned.toLowerCase()] ?? cleaned;
}
